lib/bug_helper.h

```diff
@@ -0,0 +1,142 @@
+/*
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+/* -*-c++-*- */
+
+#ifndef LIB_BUG_HELPER_H_
+#define LIB_BUG_HELPER_H_
+
+#include <string>
+#include <string_view>
+
+#include <boost/format.hpp>
+
+#include "absl/strings/str_cat.h"
+#include "lib/cstring.h"
+#include "lib/source_file.h"
+
+template <class... Args>
+std::string bug_helper(boost::format &f, std::string_view position, std::string_view tail,
+                       const char *t, Args &&...args);
+
+template <class... Args>
+std::string bug_helper(boost::format &f, std::string_view position, std::string_view tail,
+                       const cstring &t, Args &&...args);
+
+template <class... Args>
+std::string bug_helper(boost::format &f, std::string_view position, std::string_view tail,
+                       const Util::SourceInfo &info, Args &&...args);
+
+template <typename T, class... Args>
+auto bug_helper(boost::format &f, std::string_view position, std::string_view tail, const T &t,
+                Args &&...args) ->
+    typename std::enable_if_t<std::is_base_of_v<Util::IHasSourceInfo, T>, std::string>;
+
+template <typename T, class... Args>
+auto bug_helper(boost::format &f, std::string_view position, std::string_view tail, const T &t,
+                Args &&...args) ->
+    typename std::enable_if_t<!std::is_base_of_v<Util::IHasSourceInfo, T>, std::string>;
+
+template <typename T, class... Args>
+auto bug_helper(boost::format &f, std::string_view position, std::string_view tail, const T *t,
+                Args &&...args) ->
+    typename std::enable_if_t<std::is_base_of_v<Util::IHasSourceInfo, T>, std::string>;
+
+template <typename T, class... Args>
+auto bug_helper(boost::format &f, std::string_view position, std::string_view tail, const T *t,
+                Args &&...args) ->
+    typename std::enable_if_t<!std::is_base_of_v<Util::IHasSourceInfo, T>, std::string>;
+
+// actual implementations
+namespace detail {
+static inline auto getPositionTail(const Util::SourceInfo &info, std::string_view position,
+                                   std::string_view tail) {
+    std::string_view posString = info.toPositionString().string_view();
```

Comment:
I wonder if it is good idea to have the source location string representation to be a `cstring`. Presumably this won't matter much as the number of diagnostics is not that high, but still. If we turned it to `std::string` instead, we would also need to change the type here.

I understand this is not changed in this PR, I just wonder what would be the implications here if we wanted to changed `SourceInfo::toPositionString` to return `std::string`. Probably it would work as is except for the change of type here...

Reply:
There are lots of `cstring` temporaries created around diagnostics. They pollute cstring's global cache and cause extra lookups for no particular reason. In many obvious cases they could be replaces with `string_view` or `string` depending on the situation. Let's see what will left after dust will settle :)